### PR TITLE
[v4] sanitize password from filters

### PR DESCRIPTION
**Pull request by Elhebert**
### What does it do?

This PR introduce a sanitizer to the filters in the Entity Service.
All attributes having a type `password` will be removed from the filters. It also returns an error when trying to filter on an attribute that doesn't exists.

### Why is it needed?

With this we want to prevent the filtering of password fields. It could be abuse to brute-force these fields.

### CODE REVIEW

File: packages/core/utils/lib/convert-query-params.js [hunk 1]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
   throw new Error('Invalid fields parameter. Expected a string or an array of strings');
 };
 
-const convertFiltersQueryParams = filters => filters;
+const convertFiltersQueryParams = (filters, type) => {
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:221]: Adding a validator to enforce array or object seems necessary

File: packages/core/utils/lib/convert-query-params.js [hunk 2]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+
+    for (const operator in filters) {
+      const attribute = type.attributes[operator];
+
+      if (attribute) {
+        if (attribute.type === 'password') {
+          delete filters[operator];
+        }
+
+        if (attribute.type === 'relation') {
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:235]: We also have to consider components / medias & dz. Some can be filtered on some should be disabled

File: packages/core/utils/lib/convert-query-params.js [hunk 3]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+
+          if (isEmpty(filters[operator])) {
+            delete filters[operator];
+          }
+        }
+
+        continue;
+      }
+
+      if (!Array.isArray(filters[operator])) {
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:246]: this seems invalid. for $not the underlying filter can by an object

File: packages/core/utils/lib/convert-query-params.js [hunk 4]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+
+        if (['$and', '$or', '$not'].includes(key)) {
+          sanitizeFilters(filter, type);
+          return true;
+        }
+
+        const attribute = type.attributes[key];
+
+        if (!attribute) {
+          throw new Error(`You can't filter on an attribute that doens't exists`);
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:265]:
```suggestion
          throw new Error(`You can't filter on an attribute that doesn't exist`);
```

File: packages/core/utils/lib/convert-query-params.js [hunk 5]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+          if (isEmpty(filters[operator])) {
+            delete filters[operator];
+          }
+        }
+
+        continue;
+      }
+
+      if (!Array.isArray(filters[operator])) {
+        throw new Error(`You can't filter on an attribute that doens't exists`);
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:247]: the error message doesn't seem to be right here

File: packages/core/utils/lib/convert-query-params.js [hunk 6]
```diff
@@ -218,7 +218,74 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+      }
+
+      filters[operator] = filters[operator].filter(filter => {
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:250]: filtering like that will be a problem as it isn't always an array

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
1st pass of review for now.

**Review by Elhebert — COMMENTED** [review]
LGTM, but as I'm the original author and contributed to the code I can't approve the changes.

File: packages/core/utils/lib/convert-query-params.js [hunk 7]
```diff
@@ -218,7 +218,77 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
+          // Sanitize each filter
+          .map(filter => sanitizeFilters(filter, schema))
+          // Filter out empty filters
+          .filter(filter => !isObject(filter) || !isEmpty(filter))
+      );
+    }
+
+    // Here, `key` can either be an operator or an attribute name
+    for (const [key, value] of Object.entries(filters)) {
+      const removeOperator = () => delete filters[key];
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:245]: Should we avoid to mutate the filters? The other functions `convert` seem to not mutate the args

**Review by petersg83 — COMMENTED** [review]
LGTM! Just one question

**Comment by petersg83** [comment]
You said `It also returns an error when trying to filter on an attribute that doesn't exist.` I don't see where is that part. Did you remove it?

File: packages/core/utils/lib/convert-query-params.js [hunk 8]
```diff
@@ -218,7 +218,80 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:245]: you should avoid redeclaring a function in a loop when possible

File: packages/core/utils/lib/convert-query-params.js [hunk 9]
```diff
@@ -218,7 +218,80 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
 
-const convertFiltersQueryParams = filters => filters;
+const convertFiltersQueryParams = (filters, schema) => {
+  // Filters need to be either an array or an object
+  // Here we're only checking for 'object' type since typeof [] => object and typeof {} => object
+  if (!isObject(filters)) {
+    throw new Error('The filters parameter must be an object or an array');
+  }
+
+  const sanitizeFilters = (filters, schema) => {
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:228]: it doesn't look like that function needs to be redefined on each convertFiltersQUeryParams call. You could either recurse on convertFilters or put the sanitize outside

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/utils/lib/convert-query-params.js (hunk 8 above)
Comment (Convly) [reply packages/core/utils/lib/convert-query-params.js:245]: Yup, indeed

File: packages/core/utils/lib/convert-query-params.js (hunk 9 above)
Comment (Convly) [reply packages/core/utils/lib/convert-query-params.js:228]: Good catch

File: packages/core/utils/lib/convert-query-params.js (hunk 8 above)
Comment (Convly) [reply packages/core/utils/lib/convert-query-params.js]: Indeed

File: packages/core/utils/lib/convert-query-params.js [hunk 10]
```diff
@@ -218,7 +218,81 @@ const convertFieldsQueryParams = (fields, depth = 0) => {
...
-const convertFiltersQueryParams = filters => filters;
+const convertFiltersQueryParams = (filters, schema) => {
+  // Filters need to be either an array or an object
+  // Here we're only checking for 'object' type since typeof [] => object and typeof {} => object
+  if (!isObject(filters)) {
+    throw new Error('The filters parameter must be an object or an array');
+  }
+
+  // Don't mutate the original object
+  const filtersCopy = cloneDeep(filters);
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:229]: I think this might cause an issue with Dates if we call it programmatically.

Comment (Convly) [reply packages/core/utils/lib/convert-query-params.js:229]: What kind of issue do you have in mind?

From the tests I've made, it looks like dates are handled well handled.

(in this screenshot dates are stringified Date instances but in fact a new Date object is created with the value of the original one `cloneDeep(o).a.d[0].e === o.a.d[0].e` returns false) 
image

**Review by alexandrebodin — APPROVED** [review]

**Pull request merged**